Full name: Nguyễn Trọng Hiếu

Student ID: 20120083

# HW2: Parallel Execution in CUDA

**To compile your file, you can use this command:** \
`nvcc tên-file.cu -o tên-file-chạy` \
***You can use Vietnamese to anwser the questions***

## Question 1A

In [ ]:
!nvcc HW2_P1.cu -o HW2_P1 --ptxas-options=-v

ptxas info    : 0 bytes gmem
ptxas info    : Compiling entry function '_Z17reduceBlksKernel3PiS_i' for 'sm_52'
ptxas info    : Function properties for _Z17reduceBlksKernel3PiS_i
    0 bytes stack frame, 0 bytes spill stores, 0 bytes spill loads
ptxas info    : Used 10 registers, 340 bytes cmem[0]
ptxas info    : Compiling entry function '_Z17reduceBlksKernel2PiS_i' for 'sm_52'
ptxas info    : Function properties for _Z17reduceBlksKernel2PiS_i
    0 bytes stack frame, 0 bytes spill stores, 0 bytes spill loads
ptxas info    : Used 14 registers, 340 bytes cmem[0]
ptxas info    : Compiling entry function '_Z17reduceBlksKernel1PiS_i' for 'sm_52'
ptxas info    : Function properties for _Z17reduceBlksKernel1PiS_i
    0 bytes stack frame, 0 bytes spill stores, 0 bytes spill loads
ptxas info    : Used 15 registers, 340 bytes cmem[0]


In [ ]:
!./HW2_P1 512 --ptxas-options=-v

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num blocks per SM: 16
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835398144 bytes
****************************

Input size: 16777217
Host time: 43.751263 ms
correct result: 2139353559 

Kernel 1
Grid size: 8193, block size: 1024
Kernel time = 2.586208 ms
Kernel result: 2139353559 
CORRECT :)

Kernel 2
Grid size: 8193, block size: 1024
Kernel time = 2.676928 ms
Kernel result: 2139353559 
CORRECT :)

Kernel 3
Grid size: 8193, block size: 1024
Kernel time = 1.166560 ms
Kernel result: 2139353559 
CORRECT :)


In [ ]:
!nvprof ./HW2_P1

==527== NVPROF is profiling process 527, command: ./HW2_P1
**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num blocks per SM: 16
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835398144 bytes
****************************

Input size: 16777217
Host time: 43.319008 ms
correct result: 2139353559 

Kernel 1
Grid size: 8193, block size: 1024
Kernel time = 2.646112 ms
Kernel result: 2139353559 
CORRECT :)

Kernel 2
Grid size: 8193, block size: 1024
Kernel time = 2.747808 ms
Kernel result: 2139353559 
CORRECT :)

Kernel 3
Grid size: 8193, block size: 1024
Kernel time = 1.243328 ms
Kernel result: 2139353559 
CORRECT :)
==527== Profiling application: ./HW2_P1
==527== Profiling result:
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   86.84%  42.933ms         3  14.311ms  14.026ms  14.469ms  [CUDA memcpy HtoD]
                    5.45%  2.6961ms         1  2.6961ms  2.6961ms  2.6961ms  reduceBlk

### Nhận xét về thời gian thực thi các công việc trong GPU activities
**Thao tác với dữ liệu trên bộ nhớ**<br>
* _**[CUDA memcpy HtoD]**_: copy dữ liệu từ host sang device ~ Chiếm lượng thời gian lớn nhất trong quá trình thực hiện reduce.
* _**[CUDA memcpy DtoH]**_: Copy dữ liệu từ device sang host ~ Do chỉ copy 1 phần tử giá trị cuối cùng nên không tốn quá nhiều thời gian để thực thi.

**Các hàm kernel chạy trên GPU**<br>
* _**Thời gian**_: Nhanh chậm tùy thuộc vào giải thuật đưa ra.

<u>_**Nhận xét</u>**_: Đa số thời gian đều được dùng trong giao đoạn đọc ghi dữ liệu trên bộ nhớ

## Question 1B


In [ ]:
!./HW2_P1 1024

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num blocks per SM: 16
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835398144 bytes
****************************

Input size: 16777217
Host time: 46.520416 ms
correct result: 2139353559 

Kernel 1
Grid size: 8193, block size: 1024
Kernel time = 2.590432 ms
Kernel result: 2139353559 
CORRECT :)

Kernel 2
Grid size: 8193, block size: 1024
Kernel time = 2.678432 ms
Kernel result: 2139353559 
CORRECT :)

Kernel 3
Grid size: 8193, block size: 1024
Kernel time = 1.168288 ms
Kernel result: 2139353559 
CORRECT :)


In [ ]:
!./HW2_P1 512

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num blocks per SM: 16
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835398144 bytes
****************************

Input size: 16777217
Host time: 43.852383 ms
correct result: 2139353559 

Kernel 1
Grid size: 16385, block size: 512
Kernel time = 1.881216 ms
Kernel result: 2139353559 
CORRECT :)

Kernel 2
Grid size: 16385, block size: 512
Kernel time = 1.984736 ms
Kernel result: 2139353559 
CORRECT :)

Kernel 3
Grid size: 16385, block size: 512
Kernel time = 0.929056 ms
Kernel result: 2139353559 
CORRECT :)


In [ ]:
!./HW2_P1 256

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num blocks per SM: 16
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835398144 bytes
****************************

Input size: 16777217
Host time: 44.167168 ms
correct result: 2139353559 

Kernel 1
Grid size: 32769, block size: 256
Kernel time = 1.540992 ms
Kernel result: 2139353559 
CORRECT :)

Kernel 2
Grid size: 32769, block size: 256
Kernel time = 1.591776 ms
Kernel result: 2139353559 
CORRECT :)

Kernel 3
Grid size: 32769, block size: 256
Kernel time = 0.785408 ms
Kernel result: 2139353559 
CORRECT :)


In [ ]:
!./HW2_P1 128

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num blocks per SM: 16
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835398144 bytes
****************************

Input size: 16777217
Host time: 44.302368 ms
correct result: 2139353559 

Kernel 1
Grid size: 65537, block size: 128
Kernel time = 1.340928 ms
Kernel result: 2139353559 
CORRECT :)

Kernel 2
Grid size: 65537, block size: 128
Kernel time = 1.373472 ms
Kernel result: 2139353559 
CORRECT :)

Kernel 3
Grid size: 65537, block size: 128
Kernel time = 0.714592 ms
Kernel result: 2139353559 
CORRECT :)


In [ ]:
n = (1 << 24)+1;

# Get grid size
def getGridSz(n, blk_sizes = [1024, 512, 257, 128]):
    dic = {}
    for blk_sz in blk_sizes:
        dic[blk_sz] = (int((((n-1)>>1)) / blk_sz + 1))
    return dic

# Get Num blocks/SM
def getBlkPerSM(grid_sizes, blk_sizes = [1024, 512, 257, 128]):
    dic = {}
    for grid_sz, blk_sz in zip(grid_sizes, blk_sizes):
        dic[blk_sz] = (int(grid_sz/blk_sz))
    return dic

grid_szs = getGridSz(n)
blk_per_SM = getBlkPerSM(grid_szs)

print(blk_per_SM)
print(grid_szs)

{1024: 1, 512: 1, 257: 1, 128: 1}
{1024: 8193, 512: 16385, 257: 32641, 128: 65537}


Maximum number of resident blocks per SM = 16 <br>
Maximum number of resident threads per SM 1024 <br>

Block size | Grid size | Num blocks / SM | Occupancy (%) | Kernel time (ms)
--- | --- | --- | --- | ---
1024 | 8193 | 1 | 1 | 2.591584
512 | 16385 | 2 | 1 | 1.877888
256 | 32641 | 4 | 1 | 1.538240
128 | 65537 | 8 | 1 | 1.337952

#### Từ bảng biểu đã điền, bạn thử suy nghĩ và giải thích xem tại sao khi thay đổi block size thì “kernel time” lại thay đổi như vậy?
* _**tài nguyên**_ với bloksize nhỏ có thể dấn đến khả năng sử dụng các tài nguyên có sẵn trong SM tốt nhơn, từ đó giúp giảm xung đột tài nguyên và tối ưu hóa việc sử dụng bộ nhớ
 trong và bộ nhớ đệm, giúp cải thiện hiệu suất tổng thể.
* _**Warp**_ Dễ dẫn đến các wrap phân kì kếu ko kiểm soát tốt đối với giải thuật đưa ra (kernel 1).


## Question 2A

In [21]:
!nvcc HW2_P2.cu -o HW2_P2

In [22]:
!./HW2_P2

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835398144 bytes
****************************

Processing time (use host): 3942.777100 ms

Basic Matrix Multiplication:
Grid size: 32 * 32, block size: 32 * 32
Processing time (use device): 8.511712 ms
Error between device result and host result: 0.000004

Shared memory Matrix Multiplication:
Grid size: 32 * 32, block size: 32 * 32
Processing time (use device): 5.509312 ms
Error between device result and host result: 0.000004

In [23]:
!nvprof ./HW2_P2

==17410== NVPROF is profiling process 17410, command: ./HW2_P2
**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835398144 bytes
****************************

Processing time (use host): 3595.703369 ms

Basic Matrix Multiplication:
Grid size: 32 * 32, block size: 32 * 32
Processing time (use device): 8.116672 ms
Error between device result and host result: 0.000004

Shared memory Matrix Multiplication:
Grid size: 32 * 32, block size: 32 * 32
Processing time (use device): 5.278176 ms
==17410== Profiling application: ./HW2_P2
Error between device result and host result: 0.000004==17410== Profiling result:
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   34.90%  3.2693ms         1  3.2693ms  3.2693ms  3.2693ms  matrix_multiplication_kernel1(float*, float*, float*, int, int, int)
                   28.21%  2.6426ms         1  2.6426ms  2.6426ms  2.

### Nhận xét về thời gian thực thi các công việc trong GPU activities
**Thao tác với dữ liệu trên bộ nhớ**<br>
* _**[CUDA memcpy HtoD]**_: copy dữ liệu từ host sang device ~ Chiếm lượng thời gian tương đối lớn trong quá trình thực hiện reduce.
* _**[CUDA memcpy DtoH]**_: Copy dữ liệu từ device sang host ~ Do chỉ copy 1 ma trận ouput nên không tốn quá nhiều thời gian bằng việc copy từ host sang device (do host phải copy 2 ma trận qua device).

**Các hàm kernel chạy trên GPU**<br>
* _**Thời gian**_: Nhanh chậm tùy thuộc vào giải thuật đưa ra (ở đây nhắm tới việc cải tiến kernel 1 sang kernel 2 bằng cách sử dụng share memory trên mỗi SM để tăng thời gian truy cập dữ liệu).

<u>_**Nhận xét</u>**_: Đa số thời gian đều được dùng trong giao đoạn thực hiện kernel (tùy thuộc vào các kích thước của ma trận đưa vào, có thể với kích thước nhỏ sẽ tốn thời gian trong giai đoạn đọc ghi hơn là trong kernel)

## Question 2B

**For Basic Matrix Multipication**

1. How many floating operations are being performed in your matrix multiply
kernel? Explain

Để có được một phần tử của ma trận output C thì cần 2 phép toán trong n lần lặp của kernel, phép toán += và phép toán *

* _**val += A[row * n + i] * B[i * k + col]**_ <br>
Khi đó số lượng phép toán là: **gridsize.blocksize.2.n.**

2. How many global memory reads are being performed by your kernel? Explain

Đọc từ global mem 2 phần tử trong 1 lần tính toán (phần tử _**A[row * n + i]**_ và _**B[row * n + i]**_)<br>
Kernel lặp _**n**_ lần để có được giá trị val → số lần đọc là **gridsize.blocksize.2.n**


3. How many global memory writes are being performed by your kernel? Explain.

Viết giá trị tính toán được (val) xuống mảng output C 1 lần trong mỗi lần lặp của kernel (gồm _**n**_ lần lặp. Do đó số lần ghi sẽ là **gridsize.blocksize.n**.

**For Tiled Matrix Multipication**

1. How many floating operations are being performed in your matrix multiply
kernel? Explain

Ứng với n lần lặp của kernel, thực thi 2 phép toán đó là phép toán += và phép toán *

* _**val += s_A[ty][j] * s_B[j][tx]**_ <br>
Khi đó số lượng phép toán là: **2.TILE_WIDTH = 2.32 = 64**

2. How many global memory reads are being performed by your kernel? Explain

Mỗi block sẽ khởi tạo share memory và đọc dữ liệu từ global memory lên share memmory, khi đó mỗi thread trong block sẽ thực thi hành động này. Mà mỗi block sẽ đọc lên các giá trị A và B (do đó đọc 2 lần cho mỗi lần lặp) tương ứng với chiều dài và chiều rộng của mảng. Đối trường hợp chiều dài và rộng ko chia hết cho blocksz thì sẽ ko đọc và gán giá trị 0, tạm coi trường hợp gán là độc giá trị từ global lên share thì số lượng phần tử đọc lên như sau:
_**gridsize.blocksize.2.(n + TILE_WIDTH - 1) / TILE_WIDTH**_

3. How many global memory writes are being performed by your kernel? Explain

Mỗi kernel ghi duy nhất giá trị output của mảng C xuống global memory, do đó. _**Số lần ghi**_ của 1 kernel bằng số số lần gọi của 1 _**tile**_, số lượng _**tile = gridsize.blocksize.(n + TILE_WIDTH - 1) / TILE_WIDTH**_